File: Algorithm_Mentoring/25-1/cpp/8주차/줄_세우기.cpp
#include <iostream>
#include <vector>
#include <queue>
using namespace std;
using ll = long long;

const ll MAX = 50101;
const ll INF = 1e12;
ll n, m, ind[MAX];
vector <ll> adj[MAX];
queue <ll> q;

int main(){
    ios::sync_with_stdio(0); // fastio
    cin.tie(0), cout.tie(0); // fastio

    cin >> n >> m;
    while(m--){
        ll s, e; cin >> s >> e;
        adj[s].push_back(e);
        ind[e]++;
    }

    for(int i = 1;i <= n;i++){
        // 들어오는 간선이 없으면 큐에 삽입
        if(!ind[i]) q.push(i);
    }

    while(!q.empty()){
        auto cur = q.front(); q.pop();
        // 현재 큐의 값 출력
        cout << cur << " ";
        for(auto& nxt : adj[cur]){
            // 다음 정점이 들어오는 간선이 없으면 큐에 삽입
            if(!--ind[nxt]) q.push(nxt);
        }
    }

    return 0;
}

File: Algorithm_Mentoring/25-1/python/8주차/줄_세우기.py
import sys
from collections import deque
input = sys.stdin.readline

MAX = 50101
INF = 10**12

n = m = 0
ind = [0] * MAX
adj = [[] for _ in range(MAX)]
q = deque()

n, m = map(int, input().split())

for _ in range(m):
    s, e = map(int, input().split())
    adj[s].append(e)
    ind[e] += 1

for i in range(1, n + 1):
    # 들어오는 간선이 없으면 큐에 삽입
    if ind[i] == 0:
         q.append(i)

while q:
    cur = q.popleft()
    # 현재 큐의 값 출력
    print(cur, end=' ')
    for nxt in adj[cur]:
        ind[nxt] -= 1
        # 다음 정점이 들어오는 간선이 없으면 큐에 삽입
        if ind[nxt] == 0:
            q.append(nxt)

